lib/common/xml.c

```diff
@@ -30,6 +30,37 @@
 #define XML_PARSER_DEBUG 0
 #endif
 
+/*!
+ * \internal
+ * \brief Apply a function to each XML node in a tree (pre-order, depth-first)
+ *
+ * \param[in,out] xml        XML tree to traverse
+ * \param[in,out] fn         Function to call for each node (returns \c true to
+ *                           continue traversing the tree or \c false to stop)
+ * \param[in,out] user_data  Argument to \p fn
+ *
+ * \return \c true to continue traversing the tree, or \c false to stop
+ *
+ * \note This function is recursive.
+ */
+bool
+pcmk__xml_foreach_dfs(xmlNode *xml, bool (*fn)(xmlNode *, void *),
```

Comment:
This intentionally is `__xml_foreach` rather than `__xe_foreach`, although I wouldn't mind doing it the latter to consider elements only. Currently, the idea is to make the traversal as generic as possible. The `fn` can decide what to do with non-element nodes.

Another approach would be to add a `bool recursive` argument to `pcmk__xe_foreach_child()` and update the body accordingly. However, I don't like that, because that function already has four arguments and doesn't consider the root of the subtree.

Reply:
I like that. "depth-first search" is clearly tells HOW we're traversing the tree and not just THAT we're traversing the tree, which is meaningful and not just an implementation detail. But
* the "dfs" prefix isn't obvious
* pre-order DFS is usually the natural way to traverse an XML tree anyway